tingbot/button.py

```diff
@@ -1,17 +1,58 @@
+import time
+import threading
+
 from .utils import CallbackList
+from .run_loop import once
+
+event_types = ['press','long_press','down','up']
+long_click_time = 1.0
 
 class Button(object):
     def __init__(self):
-        self.callbacks = CallbackList()
+        self.callbacks = {x:CallbackList() for x in event_types}
         self.was_pressed = False
+        self.actions = []
+        self.pressed = False
+        self.down_time = 0
+        self.click_count = 0
+        self.lock = threading.Lock()
+        
+    def add_action(self,action):
+        self.lock.acquire()
+        self.actions.append(action)
+        self.lock.release()
+
+    def action(self,action):
+        if action=="down":
+            self.down_time = time.time()
+            self.click_count += 1
+            self.pressed_click_count = self.click_count
+            once(seconds=long_click_time)(lambda: self.long_press(self.click_count))
```

Comment:
I'm not sure we need to use the main thread run loop here. The main thread run loop isn't thread safe (I'm not sure it needs to be for normal usage), so this could cause problems there, and the ordering of this timer relative to the button callback isn't known, so the ordering of actions could change depending on the run loop actions.